nemo_aligner/models/nlp/gpt/megatron_gpt_spin_model.py

```diff
@@ -400,18 +397,95 @@ def onload_adam_states(self):
-
-    def sharded_state_dict(self, prefix: str = ""):
-        sharded_state_dict = super().sharded_state_dict(prefix=prefix)
-
-        # add in the reference policy weights
+    
+    # Alternative 1: swap in the ref weights and create a custom key which then needs to be handled in the loader
+    def sharded_state_dict_alt_1(self, prefix: str = ''):
+        sharded_state_dict_orig = super().sharded_state_dict(prefix=prefix)
+        
+        if self.ref_policy_state_dict is None:
+            return sharded_state_dict_orig
+        
+        sharded_state_dict_new = {}
+        with cpu_weight_swap(self, self.ref_policy_state_dict, megatron_amp_O2=self.megatron_amp_O2):
+            module_prefix = f'{prefix}ref_policy_model.'
+            for index, module in enumerate(self.get_gpt_module_list()):
+                if parallel_state.get_virtual_pipeline_model_parallel_world_size() is not None:
+                    # virtual pipline rank must be set so that GPTModel returns the correct sharded state dict
+                    parallel_state.set_virtual_pipeline_model_parallel_rank(index)
+                    module_sharded_state_dict = module.sharded_state_dict(prefix=module_prefix)
+                    sharded_state_dict_new[f'ref_policy_model_{index}'] = module_sharded_state_dict
+                else:
+                    module_sharded_state_dict = module.sharded_state_dict(prefix=module_prefix)
+                    sharded_state_dict_new.update(module_sharded_state_dict)
+
+            # reset vp rank
+            if parallel_state.get_virtual_pipeline_model_parallel_world_size() is not None:
+                parallel_state.set_virtual_pipeline_model_parallel_rank(0)
+        
+        return sharded_state_dict_orig | sharded_state_dict_new
+    
+    # Alternative 2: use make_sharded_optimizer_tensor to link the TP/PP/DP info, but this may break under VP logic
+    def sharded_state_dict(self, prefix: str = ''):
+        sharded_state_dict_orig = super().sharded_state_dict(prefix=prefix)
+        
         if self.ref_policy_state_dict is not None:
-            sharded_state_dict["reference_policy"] = make_sharded_tensors_for_checkpoint(
-                self.ref_policy_state_dict, prefix
-            )
+            # Link ref_policy keys with sharded_state_dict to reuse sharding information
+            ref_policy_sharded_state_dict = {}
+            for k in self.ref_policy_state_dict:
+                key = k.replace("model.module.", "model.", 1) if self.megatron_amp_O2 else k
+                assert key in sharded_state_dict_orig, f"key [ {key} ] exists in ref_policy but not in sharded_state_dict_orig" # may fail due to nesting?
+                ref_policy_sharded_state_dict[k] = make_sharded_optimizer_tensor_and_state(sharded_state_dict_orig[key], self.ref_policy_state_dict[k], 'reference_policy')
+            sharded_state_dict_orig['reference_policy'] = ref_policy_sharded_state_dict
```

Comment:
Related to your comment above about nesting:

I wonder if there's a risk of some code breaking (in the future) when we use nested sharded state dicts. Would it be safer to keep it flat and instead rely on the `reference_policy` prefix to (i) prevent overlap with current params, and (ii) reconstruct the ref policy state dict in `on_load_checkpoint()`?

Reply:
Moving to nesting may indeed cause several breaking changes, not just here, but with loads of other code in Aligner (for example `retrieve_model_state_dict_in_cpu`. I think if we move to nested state_dicts we will need to change a lot and run several tests. Ideally, everything would remain flat, but that unfortunately isn't possible if using VP

Reply:
PS: I just had a quick look at your `alt_1` implementation (which I had skipped earlier as I wasn't sure it was supposed to work), and actually what I describe (in terms of keeping the dict flat) seems to be exactly what you did there.

Reply:
`f"{prefix}reference_policy"`

This is already happening inside of `utils.make_sharded_tensors_from_reference`. If you look at the code of that function, and the fact that I pass in "reference_policy" you will see that the key it uses is what you propose.

Using Alt_2 as it is now, I can see that everything is saved down correctly when I look in the root of the checkpoint directory:

```
drwxr-xr-x  2 dip dip 4.0K Feb 29 19:00 model.decoder.final_layernorm.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 18:59 model.decoder.layers.mlp.linear_fc1._extra_state
drwxr-xr-x  2 dip dip 4.0K Feb 29 19:00 model.decoder.layers.mlp.linear_fc1.layer_norm_weight
drwxr-xr-x  2 dip dip  20K Feb 29 19:00 model.decoder.layers.mlp.linear_fc1.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 18:59 model.decoder.layers.mlp.linear_fc2._extra_state
drwxr-xr-x  2 dip dip  12K Feb 29 19:00 model.decoder.layers.mlp.linear_fc2.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 18:59 model.decoder.layers.self_attention.linear_proj._extra_state
drwxr-xr-x  2 dip dip  12K Feb 29 19:00 model.decoder.layers.self_attention.linear_proj.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 18:59 model.decoder.layers.self_attention.linear_qkv._extra_state
drwxr-xr-x  2 dip dip 4.0K Feb 29 19:00 model.decoder.layers.self_attention.linear_qkv.layer_norm_weight
drwxr-xr-x  2 dip dip  12K Feb 29 19:00 model.decoder.layers.self_attention.linear_qkv.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 18:59 model.embedding.word_embeddings.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 19:00 model.output_layer.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 18:59 optimizer.state.exp_avg.model.decoder.final_layernorm.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 18:59 optimizer.state.exp_avg.model.decoder.layers.mlp.linear_fc1.layer_norm_weight
drwxr-xr-x  2 dip dip  20K Feb 29 18:59 optimizer.state.exp_avg.model.decoder.layers.mlp.linear_fc1.weight
drwxr-xr-x  2 dip dip  12K Feb 29 18:59 optimizer.state.exp_avg.model.decoder.layers.mlp.linear_fc2.weight
drwxr-xr-x  2 dip dip  12K Feb 29 18:59 optimizer.state.exp_avg.model.decoder.layers.self_attention.linear_proj.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 18:59 optimizer.state.exp_avg.model.decoder.layers.self_attention.linear_qkv.layer_norm_weight
drwxr-xr-x  2 dip dip  12K Feb 29 18:59 optimizer.state.exp_avg.model.decoder.layers.self_attention.linear_qkv.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 18:59 optimizer.state.exp_avg.model.embedding.word_embeddings.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 18:59 optimizer.state.exp_avg.model.output_layer.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 18:59 optimizer.state.exp_avg_sq.model.decoder.final_layernorm.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 18:59 optimizer.state.exp_avg_sq.model.decoder.layers.mlp.linear_fc1.layer_norm_weight
drwxr-xr-x  2 dip dip  16K Feb 29 18:59 optimizer.state.exp_avg_sq.model.decoder.layers.mlp.linear_fc1.weight
drwxr-xr-x  2 dip dip  12K Feb 29 18:59 optimizer.state.exp_avg_sq.model.decoder.layers.mlp.linear_fc2.weight
drwxr-xr-x  2 dip dip  12K Feb 29 18:59 optimizer.state.exp_avg_sq.model.decoder.layers.self_attention.linear_proj.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 18:59 optimizer.state.exp_avg_sq.model.decoder.layers.self_attention.linear_qkv.layer_norm_weight
drwxr-xr-x  2 dip dip  12K Feb 29 18:59 optimizer.state.exp_avg_sq.model.decoder.layers.self_attention.linear_qkv.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 18:59 optimizer.state.exp_avg_sq.model.embedding.word_embeddings.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 18:59 optimizer.state.exp_avg_sq.model.output_layer.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 18:59 optimizer.state.fp32_param.model.decoder.final_layernorm.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 18:59 optimizer.state.fp32_param.model.decoder.layers.mlp.linear_fc1.layer_norm_weight
drwxr-xr-x  2 dip dip  12K Feb 29 18:59 optimizer.state.fp32_param.model.decoder.layers.mlp.linear_fc2.weight
drwxr-xr-x  2 dip dip  12K Feb 29 18:59 optimizer.state.fp32_param.model.decoder.layers.self_attention.linear_proj.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 18:59 optimizer.state.fp32_param.model.decoder.layers.self_attention.linear_qkv.layer_norm_weight
drwxr-xr-x  2 dip dip  12K Feb 29 18:59 optimizer.state.fp32_param.model.decoder.layers.self_attention.linear_qkv.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 18:59 optimizer.state.fp32_param.model.embedding.word_embeddings.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 18:59 optimizer.state.fp32_param.model.output_layer.weight
drwxr-xr-x  2 dip dip  16K Feb 29 18:59 optimizer.state.param.model.decoder.layers.mlp.linear_fc1.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 19:00 reference_policy.model.decoder.final_layernorm.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 18:59 reference_policy.model.decoder.layers.mlp.linear_fc1._extra_state
drwxr-xr-x  2 dip dip 4.0K Feb 29 19:00 reference_policy.model.decoder.layers.mlp.linear_fc1.layer_norm_weight
drwxr-xr-x  2 dip dip  20K Feb 29 19:00 reference_policy.model.decoder.layers.mlp.linear_fc1.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 18:59 reference_policy.model.decoder.layers.mlp.linear_fc2._extra_state
drwxr-xr-x  2 dip dip  12K Feb 29 19:00 reference_policy.model.decoder.layers.mlp.linear_fc2.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 18:59 reference_policy.model.decoder.layers.self_attention.linear_proj._extra_state
drwxr-xr-x  2 dip dip  12K Feb 29 19:00 reference_policy.model.decoder.layers.self_attention.linear_proj.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 18:59 reference_policy.model.decoder.layers.self_attention.linear_qkv._extra_state
drwxr-xr-x  2 dip dip 4.0K Feb 29 19:00 reference_policy.model.decoder.layers.self_attention.linear_qkv.layer_norm_weight
drwxr-xr-x  2 dip dip  12K Feb 29 19:00 reference_policy.model.decoder.layers.self_attention.linear_qkv.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 19:00 reference_policy.model.embedding.word_embeddings.weight
drwxr-xr-x  2 dip dip 4.0K Feb 29 19:00 reference_policy.model.output_layer.weight
```

I understand what you mean by flat now though, and I'm honestly not sure if that is necessary. I think it's okay to leave it for now and then engineer for confirmed cases in the future rather than trying to over-engineer now for possible future cases which may never come to fruition.

Reply:
Ok, I'm not familiar enough with this distributed checkpointing code to know for sure what is the "right" way of doing things, so let's keep your current implementation for now since it works.